flow terminology (#224)
cylc/cylc-doc: src/glossary.rst

.. glossary::
   :sorted:

   workflow
   cylc workflow
      A Cylc workflow is a collection of :term:`tasks` to carry out and
      :term:`dependencies` that govern the order in which they run. This is

>       represented in Cylc format in a :cylc:conf:`flow.cylc` file.

Reviewer: How about a note that says workflows used to be called suites?

      * :term:`run directory`

   source directory
      Any directory where :term:`workflows <workflow>` are written and stored
      in preparation for installation with ``cylc install`` or reinstallation

>       with ``cylc reinstall``.

Reviewer: Might be worth mentioning the search path can be configured in `global.cylc[install]source dirs`

      See also:

      * :term:`run directory`
      * :ref:`Installing-workflows`

   run directory
      This is a directory containing the configuration that Cylc uses to run
      the :term:`workflow`.

      Typically this is installed from the :term:`source directory` using
      ``cylc install``.

      The run directory can be accessed by a running workflow using
      the environment variable ``CYLC_SUITE_RUN_DIR``.

      See also:

      * :term:`source directory`
      * :term:`work directory`
      * :term:`share directory`
      * :term:`job log directory`
      * :term:`cylc-run directory`

   play
      We run a :term:`workflow` using the ``cylc play`` command.

      You can :term:`play`, :term:`pause` and :term:`stop` a :term:`flow`,
      Cylc will always carry on where it left off.

      See also:

      * :term:`pause`
      * :term:`stop`

>    pause

Reviewer: just realised, there's already a definition for pause on master, is this branch a bit behind?

      See also:

      * :ref:`Graph Branching`

source workflow
   A source workflow is a :term:`cylc workflow` that has not yet been

>    :term:`installed` into the :term:`run directory`.

Reviewer: Only thing is, the source workflow still exists separately after it has been installed.

Author: Yep, on second read this isn't very good, attempt two.

      :caption: flow.cylc

      [scheduling]
          cycling mode = integer
          initial cycle point = 1
          [[graph]]
              # repeat this for each batch
              P1 = """
                  # the stages of brewing in the order they must occur in
                  malt => mash => sparge => boil => chill => ferment => rack

                  # must finish the sparge of one batch before
                  # starting the next one
                  # sparge[-P1] => mash
              """

.. TODO - add this after the universal ID work

   workflow id
      A workflow ID is the name Cylc uses to identify a :term:`workflow` in
      the :term:`run directory`.

      This ID is used on the command line and in the GUI.

      The ID is the relative path between the :term:`workflows <workflow>`
      :term:`run directory` and the :term:`cylc run directory`.

      For example, if your workflow is in ``~/cylc-run/foo/bar/run1`` then its ID
      is ``foo/bar/run1``.

      .. ..note:: For convenience you can omit the ``/run1`` on the command line.

flow
   A flow is a single logical run of a :term:`workflow` that is done by
   a :term:`scheduler`.

>    A flow can be :term:`played <play>` and :term:`paused <pause>`,

Reviewer: I don't think there is a defintion for "play" at the moment, I suppose it's "causing the scheduler to run, either by starting it, restarting it or resuming it"

Author: I think there are a few definitions missing ATM, waiting to get the "flow" stuff sorted before filling in the blanks.

Reviewer 2: Would these emoji's potentially make our conceit RE: The control terminology more obvious? ⏯️ ⏹️

>    :term:`started <start>` and :term:`stopped <stop>`, it is still the same
>    flow.

Reviewer: ```suggestion
   :term:`started <start>` and :term:`stopped <stop>`, and will remain
   the same flow.
```

      See also:

      * :ref:`Graph Branching`

> source workflow

Reviewer 2: workflow source?

Author: The source workflow is in the workflow source.

Reviewer 2: If you can't play it then surely it's not a workflow?

Reviewer: Middle ground: workflow source workflow?

Reviewer 3: Programming analogies might help...

IMO "source code" refers generically to the medium. i.e., that bunch of text is "source code".  It doesn't mean, that is the source for this particular computer program. For that, you would say "program source", not "source program" which sounds like "a program that is the source of something else", not "the source code of the program".

So on that basis, I think we should go with:

*The **workflow source directory" contains the workflow config file that gets installed to (and is therefore the source of) this workflow run directory.*

We don't necessarily have to confuse people by refering to the workflow config file itself as a "workflow source" or "source workflow".

   A source workflow is a :term:`cylc workflow` that is not located in the
   :term:`cylc run directory`.

   In order to be :term:`run` source workflows must be
   :term:`installed <install>` into the :term:`cylc run directory`.

workflow
cylc workflow

>    A Cylc workflow is a collection of things to run (:term:`tasks`) along with

Reviewer 2: "things" → "tasks"

   the constraints that govern the order in which they run
   (:term:`dependencies`) represented in Cylc format in a
   :cylc:conf:`flow.cylc` file.

>    For example here is a Cylc workflow representing the brewing process:

Reviewer 2: 🤣 How very you....

Reviewer 2: Perhaps we should jot some abstract examples down on cylc-admin:
Off the top of my head:
- Cooking
- Gardening 
- DIY
- Start of Driving lesson (chair, seatbelt, mirrors, hands at 10 and 2, clutch (if manual), choke(if old car), start, mirror, signal, maneuver, possible branch )
-

Reviewer 3: > a disturbingly small proportion of the population are familiar with the brewing process

But a gratifying large proportion are familiar with the products thereof :beers:

flow
   A flow is a single logical run of a :term:`workflow` that is done by
   a :term:`scheduler`.

   A flow can be :term:`played <play>` and :term:`paused <pause>`,
   :term:`stopped <stop>` and :term:`restarted <restart>`.

   A flow begins at the :term:`start cycle point` and ends at the
   :term:`stop cycle point`.

   It is possible to run more than one flow in a single :term:`scheduler`.

reflow
   A reflow is a subsequent logical run of a :term:`workflow` that is done by
   the same :term:`scheduler` as the original :term:`flow`.

   Reflows are useful when you need to re-wind your :term:`workflow` run to
   allow it to evolve a new path into the future.

   .. TODO - add this once the CLI example works

      For example, the following workflow runs the task ``a`` every cycle, then
      either ``b`` or ``c`` depending on the outcome of ``a``:

      .. code-block:: cylc

         [scheduling]
             cycling mode = integer
             initial cycle point = 1
             [[graph]]
                 P1 = """
                     b[-P1] | c[-P1] => a
                     a:succeed => b
                     a:fail => c

>                  """

Reviewer 3: We'll need to start adding `allow implicit tasks = True` to all our example workflows so that they can run out of the box.

Author: Ach, dammit, that makes examples so convoluted.

Reviewer 2: or define the tasks in the runtime section?

Author: Even worse?

      Let's say the workflow has run to cycle 8, but we have just noticed that
      something went wrong with the task ``a`` in cycle 5.

      To rectify this we could create a new flow (a reflow) starting at ``a.5``::

         # get the ID of the original flow

>          cylc TBC <id> a.5  # ???

Reviewer 3: At the moment, have to get it from the log:
>  (a.10) spawned b.10 flow(z)

         # pause the original flow

>          cylc pause <id> --flow <flow>

Reviewer 3: We can't do this yet.  We need to decide on scheduler pause vs task hold implementation. Probably the latter, i.e. we'd end up with held tasks from the targeted flow, in which case this example should use the `cylc hold` command.

Reviewer 3: We can already `cylc stop` a particular flow, in which case current active tasks in the flow will not spawn downstream children.

Reviewer 3: Also, holding tasks is currently problematic: cylc/cylc-flow#3743

Author: If `cylc hold '*'` targets all tasks in the n=1 window then that would do the trick.

(off topic for this PR, n=1 matching + a new list in the task_pool caching hold requests for proxies that aren't in the pool yet backed up by a new DB table should do the trick)

Author: (resolving this as the code is a comment that is not rendered, we can fix this once we have reflow sorted)

         # create a new flow starting at a.5
         cylc trigger a.5 --reflow

      Where ``<id>`` is the name of the :term:`workflow` and ``<flow>`` is the
      identifier of the :term:`flow`.

>       This new flow will run over the top of the old one.

Reviewer 3: If the old flow was *stopped*, the new flow will carry on past the point where the old flow stopped.

If the old flow was *held*(/(*paused*?), the new flow will run into the held tasks of the new flow, and the flow IDs will merge there.

Author: Explicitly stopping is probably un-necessary, however, a bit cleaner?

If the old and new flows take different paths a `cylc release '*'` could potentially result in the old flow continuing?

Reviewer 3: > Explicitly stopping is probably un-necessary, however, a bit cleaner?

Yes, but easier to achieve, hence we have it already :grin: 

> If the old and new flows take different paths a cylc release '*' could potentially result in the old flow continuing?

Yes (unless you "stop" it)
